bootstore/src/schemes/mod.rs

```diff
@@ -0,0 +1,101 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Versioned bootstore protocol schemes
+//!
+//! Contains code shared by all schemes
+
+use serde::{Deserialize, Serialize};
+
+pub mod v0;
+
+// We keep these in a module to prevent naming conflicts
+#[allow(unused)]
+pub mod params {
+    #[derive(Default, Debug, Clone, Copy)]
+    pub struct ChaCha20Poly1305;
+
+    #[derive(Default, Debug, Clone, Copy)]
+    pub struct Sha3_256;
+
+    #[derive(Default, Debug, Clone, Copy)]
+    pub struct Hkdf;
+
+    #[derive(Default, Debug, Clone, Copy)]
+    pub struct Tcp;
+
+    #[derive(Default, Debug, Clone, Copy)]
+    pub struct No;
+
+    #[derive(Default, Debug, Clone, Copy)]
+    pub struct Curve25519;
+
+    #[derive(Default, Debug, Clone, Copy)]
+    pub struct Bcs;
+
+    #[derive(Default, Debug, Clone, Copy)]
+    pub struct U32BigEndian;
+}
+
+/// A common message supported across all schemes.
+///
+/// The Hello message must remain backwards compatible for all time, and so it
+/// is serialized as two big-endian 4-byte integers, followed by 1 8 byte big-
+/// endian reserved integer. This allows serialization format to change across
+/// schemes but not break version negotiation. The use of big-endian integers
+/// allows easier reading of wire dumps.
+///
+/// At the initiation of each session, each side sends a `Hello` message
+/// indicating the current scheme and version they are operating at.
```

Comment:
It would be helpful to clarify what is expected to change in different schemes vs versions here.  I'm assuming this is like the [`transceiver-control` protocol](https://github.com/oxidecomputer/transceiver-control/blob/main/messages/src/message.rs#L26), where versions are always backwards-compatible (but couldn't be able to deserialize newer messages).